### PR TITLE
[FEATURE] New "Merged feature" renderer for polygon and line layers

### LINKED ISSUE #21642: Merge connected lines option for Marker line interval
Author Name: **Matt Sim** (Matt Sim)
Original Redmine Issue: [13600](https://issues.qgis.org/issues/13600)

Redmine category:symbology

---
Similar to the label rending option "Merge connected lines to avoid duplicate labels". It would be great if the Marker lines could be merged, allowing the marker placement interval to be consistent across connected features.

### LINKED ISSUE #27540: Symbology option: suppress shared polygon boundaries
Author Name: **Nathan Perry** (Nathan Perry)
Original Redmine Issue: [19715](https://issues.qgis.org/issues/19715)

Redmine category:symbology

---
It would be tremendously useful if the Symbology panel had a built-in option to style polygon features as if they were dissolved; that is, rendering them as a single united polygon with no internal boundaries visible. Currently this is only possible using the collect() function in Geometry Generator, or as a workaround with multiple symbol layers and offset outlines, but is commonly used enough to warrant a core feature.

This could be implemented as a check box in the Symbol options for fill-type symbols, perhaps: "Suppress shared polygon boundaries"

### PR DESCRIPTION
**Pull request by nyalldawson**
This renderer merges (or unions/dissolves) the line or polygon features
from a layer prior to rendering them. It's useful for rendering
a polygon layer with overlapping features as one single "coverage" feature,
or a line layer consisting of many smaller component line features
using a regularly spaced marker line or similar.

Internally, this just moves the guts of the existing inverted polygons
renderer to a new base class, as that renderer already had an option
to merge features prior to rendering. Basically it just exposes a new
renderer to users which is the inverted polygon renderer without
the inversion step!

Using single symbol renderer, with marker line symbol:

image

Using merged feature renderer:

image

Fixes #21642
Fixes #27540

### CODE REVIEW

**Comment by pjshelton** [comment]
Fixes #27540

**Comment by pjshelton** [comment]
Does this have the ability to group by attribute?  So only merge if an expression/attribute has the same value? 
If it does it would also close #14953.

It also closes #27540 which is the polygon version of #21642

@nyalldawson I don't know how and am not sure I would have the right permissions to create the required links.

**Comment by nyalldawson** [comment]
It maintains the renderer classes, so if you use a categorised subrenderer then only features from the same class are merged

**Comment by pjshelton** [comment]
image

What is the expected behaviour of the marker line when the Merged linestring is a Multilinestring?

If all the lines in the attached image get merged to one object, how is the spacing on the "Y"s going to get handled?

**Comment by pjshelton** [comment]
And will it work with a generated geometry?  If my layer is a polygon layer, and I want to put a marker symbol on the boundary, but don't want the common boundary double marked (see #17943) I don't want the polygons merged, just their boundary.  I am hoping that one could use a generated geometry to render the polygon as a linestring, and then use this feature to merge the common boundaries.

**Comment by nyalldawson** [comment]
> If all the lines in the attached image get merged to one object, how is the spacing on the "Y"s going to get handled?

It's unpredictable - depending on the feature order you may get either the all the "main line" sections merged OR a bit of the main line + one of the branches

> And will it work with a generated geometry? If my layer is a polygon layer, and I want to put a marker symbol on the boundary, but don't want the common boundary double marked (see #17943) I don't want the polygons merged, just their boundary. I am hoping that one could use a generated geometry to render the polygon as a linestring, and then use this feature to merge the common boundaries.

That's a non-trivial use case not yet handled by this renderer, sorry

File: src/core/symbology/qgsmergedfeaturerenderer.h [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    void setLegendSymbolItem( const QString &key, QgsSymbol *symbol ) override;
+    bool legendSymbolItemsCheckable() const override;
+    bool legendSymbolItemChecked( const QString &key ) override;
+    void checkLegendSymbolItem( const QString &key, bool state = true ) override;
+    bool accept( QgsStyleEntityVisitorInterface *visitor ) const override;
+
+    /**
+     * Creates a QgsMergedFeatureRenderer by a conversion from an existing renderer.
+     * \returns a new renderer if the conversion was possible, otherwise 0.
+     * \since QGIS 2.5
```
Comment (nirvn) [inline src/core/symbology/qgsmergedfeaturerenderer.h:102]: Take the since 2.5 out?

**Review by nirvn — APPROVED** [review]
Fantastic, it'll come in very handy.

**Comment by gioman** [comment]
I was kind of hoping that this would have helped with the representation of adjacent polygon boundaries with marker lines (for example), but I guess it is not the case, or am I using it the wrong way?

**Comment by nyalldawson** [comment]
@gioman 
>  I was kind of hoping that this would have helped with the representation of adjacent polygon boundaries with marker lines (for example), but I guess it is not the case, or am I using it the wrong way?

No, you're not missing anything -- it's not supported by the renderer.

We could potentially make this an option in the renderer, basically by taking the boundary of the input polygons and then merging those. But I suspect the result won't be very usable in most situations as it would only work for exactly coincidental borders. What we'd really need in order to handle this would be a way to have some tolerance in the border collapsing -- e.g. if they differ by < 1 mm then treat them as identical/collapsible. That's not a trivial change, unfortunately.

**Comment by gioman** [comment]
> We could potentially make this an option in the renderer, basically by taking the boundary of the input polygons and then merging those. But I suspect the result won't be very usable in most situations as it would only work for exactly coincidental borders. What we'd really need in order to handle this would be a way to have some tolerance in the border collapsing -- e.g. if they differ by < 1 mm then treat them as identical/collapsible. That's not a trivial change, unfortunately.

Thanks for clarifying!

**Pull request merged**